pkg/pillar/cmd/usbmanager/usbevent.go

```diff
@@ -11,72 +11,115 @@ import (
 	"github.com/eshard/uevent"
 )
 
+type ueventListener struct {
+	isCancelled  atomic.Bool
+	ueventReader io.ReadCloser
+	usbdevices   map[string]*usbdevice
+	uc           *usbmanagerController
+}
+
 func (uc *usbmanagerController) listenUSBPorts() {
-	var isCancelled atomic.Bool
+	ul := ueventListener{
+		usbdevices: map[string]*usbdevice{},
+		uc:         uc,
+	}
+	ul.isCancelled.Store(false)
+	uc.listenUSBStopChan = make(chan struct{})
 
-	usbdevices := make(map[string]*usbdevice)
-	isCancelled.Store(false)
+	ul.waitForStop()
 
-	r, err := uevent.NewReader()
-	if err != nil {
-		log.Fatal(err)
-	}
+	go func() {
+		retry := true
+		for retry {
+			retry = ul.readEvents()
+		}
+	}()
+}
 
+func (ul *ueventListener) waitForStop() {
 	// goroutine for cancelling uevent reader
-	uc.listenUSBStopChan = make(chan struct{})
 	go func() {
-		<-uc.listenUSBStopChan
-		isCancelled.Store(true)
-		r.Close()
+		<-ul.uc.listenUSBStopChan
+		ul.isCancelled.Store(true)
+		if ul.ueventReader != nil {
+			ul.ueventReader.Close()
+		}
 		return
 	}()
+}
 
-	go func() {
-		for _, ud := range walkUSBPorts() {
-			usbdevices[ud.ueventFilePath] = ud
-			log.Tracef("usb device from walking: %+v", ud)
-			uc.addUSBDevice(*ud)
-		}
+func (ul *ueventListener) readEvents() bool {
+	var err error
+	ul.ueventReader, err = uevent.NewReader()
+	if err != nil {
+		log.Warnf("Opening uevent reader failed: %v - retrying", err)
```

Comment:
Did you intent to return true here as well?

Reply:
Yes, definitely it should.